docs/testing.md

```diff
@@ -1,13 +1,89 @@
-
-## Utilities
-
-### Unit testing
-
-#### `createMockExecutionContext`
+# Testing Integrations
+
+A JupiterOne integration is expected to produce a graph of entities and
+relationships representing the state of something an organization cares about. A
+well factored integration is typically composed of three major components:
+
+1. Data source clients authenticate with and collect data from the source
+1. Converters transfer source data to graph entity and relationship properties
+1. Integration steps combine collection and conversion to build a subset of the
+   graph
+
+Data collection code encapsulates authentication and provides functions such as
+`iterateUsers`, `iterateGroups`, and `iterateRepositories`, which do nothing
+more than paginate the proper endpoint(s) to iterate each major asset of the
+source data, and encapsulate response/data loading error handling. This single
+responsibility provides an opportunity to extract the code into a client for
+general use against the targeted API/data source. The iterator approach should
+drive clients to avoid loading complete sets of provider data into memory.
+
+Converter functions take source data and produce an entity or relationship graph
+object. These typically wrap up a call to `createIntegrationEntity` or
+`createIntegrationRelationship`, which can handle most of the work. However, it
+is important to write a test that expresses the expected properties of graph
+objects so that the integration remains stable throughout changes to these
+helper functions.
+
+Integration steps reflect an understanding of how source data is related and the
+order in which data must be obtained. This step architecture allows for
+concurrent execution of fetching data that is not related, and sequential
+execution of fetching data that depends on other data having already been
+fetched. Steps iterate source objects using the data client and converters to
+produce entities and relationships, but the steps are responsible for organizing
+data collection and entity and relationship construction.
+
+## Integration Testing
+
+Ideal integration testing (ensuring end-to-end success) would:
+
+1. Consistently create data in the source system
+1. Execute the integration
+1. Verify the expected graph was produced
+
+This can be accomplished by:
+
+1. Scripting source data setup
+1. Executing the integration
+1. Describing the complete set of entities and relationships
+
+We are looking at ways to make this easy.
+
+## Unit Testing
+
+Strive to isolate unit tests according to:
+
+1. Client authentication and data fetching, including response error handling
+1. Converting source data to entity/relationship objects, ensuring properties
+   are transferred as expected
+1. Producing entities with the right relationships based on a known set of
+   source data (do not test properties)
+
+Use [Polly.js](https://netflix.github.io/pollyjs) for testing the client code.
```

Comment:
I would specify making this where applicable.  With the snowflake SDK it makes weird requests (pre flight exec, post flight exec and return data in difficult formats, also pretty sure some of it is over raw TCP) and the return data is hard to figure out what is what.

Reply:
Okay, will do!